### PR TITLE
Add grammar for cfn-guard rulesets

### PR DESCRIPTION
**Pull request by jotompki**
*Issue #, if available:*

*Description of changes:* [ABNF](https://en.wikipedia.org/wiki/Augmented_Backus%E2%80%93Naur_form) definition of the cfn-guard ruleset grammar. Looked at the [cfn-guard README](https://github.com/aws-cloudformation/cloudformation-guard/blob/master/cfn-guard/README.md) and the [parser](https://github.com/aws-cloudformation/cloudformation-guard/blob/master/cfn-guard/src/parser.rs) to determine some of this. Still a work in progress so leaving as a draft for now to hone and collect feedback.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by PatMyron** [comment]
Should [environment variables](https://github.com/aws-cloudformation/cloudformation-guard/tree/master/cfn-guard#environment-variables) be added?

**Comment by jotompki** [comment]
> Should [environment variables](https://github.com/aws-cloudformation/cloudformation-guard/tree/master/cfn-guard#environment-variables) be added?

Good call. Will add that as a terminal case

File: GRAMMAR.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+<RULE_LINE> ::= <RULE> | <RULE> <BOOL_OPERAND> <RULE_LINE>
+<RULE> ::= <BASE_RULE> <OUTPUT_MESSAGE> | <BASE_RULE>
+<BASE_RULE> ::= <CHECK_RULE> | <CONDITIONAL_RULE>
+<CHECK_RULE> ::= <RESOURCE_TYPE> <PROPERTY_CHECK>
+<CHECK_VALUE> ::= <PRIMITIVE> | %<VARIABLE>
+<PROPERTY_CHECK> ::= <PROPERTY_PATH> <OPERAND> <CHECK_VALUE>
+<CONDITIONAL_RULE> ::= <RESOURCE_TYPE> WHEN <PROPERTY_CHECK> CHECK <PROPERTY_CHECK>
+<VARIABLE> ::= define variable
+<OPERAND> ::= ==|!=|<|>|<=|>=|IN|NOT_IN
+<BOOL_OPERAND> ::= |OR| | |AND|
```
Comment (PatMyron) [inline GRAMMAR.md:16]: under the impression an explicit `|AND|` operator didn't exist yet

Comment (jotompki) [reply GRAMMAR.md:16]: [There's some logic in place for it currently](https://github.com/aws-cloudformation/cloudformation-guard/blob/master/cfn-guard/src/parser.rs#L243). I'll verify that if works.

**Review by ammokhov — COMMENTED** [review]
this looks good if you are familiar with semantics but should we do a resource example?

**Comment by jotompki** [comment]
> this looks good if you are familiar with semantics but should we do a resource example?

Resource example? like an example of a ruleset? the readme has several examples of valid rules: https://github.com/aws-cloudformation/cloudformation-guard/blob/master/README.md Do you think we should include more here? Was just shooting to scope this document to defining the grammar.

**Review by ammokhov — APPROVED** [review]
> > this looks good if you are familiar with semantics but should we do a resource example?
> 
> Resource example? like an example of a ruleset? the readme has several examples of valid rules: https://github.com/aws-cloudformation/cloudformation-guard/blob/master/README.md Do you think we should include more here? Was just shooting to scope this document to defining the grammar.

nice abnf addition, lgtm

File: GRAMMAR.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
+# Overview
+Rulesets can be described by the following ABNF grammar:
+
+```
```
Comment (PatMyron) [inline GRAMMAR.md:4]:
```suggestion
```abnf
```

should be more readable with [ABNF](https://github.com/github/linguist/blob/46b583d16b80a7a12896a1e6ca16ea345e8e7491/lib/linguist/languages.yml#L64) [syntax-highlighting](https://docs.github.com/en/free-pro-team@latest/github/writing-on-github/creating-and-highlighting-code-blocks#syntax-highlighting)

File: GRAMMAR.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+regex = "/" vchar-sp "/"; regular expression in rust regex syntax: https://docs.rs/regex/1.3.9/regex/#syntax
+csv = csv-value *(value-separator csv-value); if json array is not valid, cfn-guard will split by commas to make a list (elements can be null)
+csv-value = [unquoted-string]
+unquoted-string = VCHAR vchar-sp; unquoted string used in the RHS of cfn-guard assignments and equality comparisons.
+
+vchar-sp = *(VCHAR / WSP)
+```
+
+References to `value` refer to the `value` from [JSON's ABNF](https://trac.ietf.org/trac/json/browser/abnf/json.abnf?rev=2) from [RFC7159](https://tools.ietf.org/html/rfc7159), which is provided below for convenience:
+```
```
Comment (PatMyron) [inline GRAMMAR.md:51]:
```suggestion
```abnf
```

should be more readable with [ABNF](https://github.com/github/linguist/blob/46b583d16b80a7a12896a1e6ca16ea345e8e7491/lib/linguist/languages.yml#L64) [syntax-highlighting](https://docs.github.com/en/free-pro-team@latest/github/writing-on-github/creating-and-highlighting-code-blocks#syntax-highlighting)

**Review by PatMyron — CHANGES_REQUESTED** [review]

**Pull request merged**